src/scenic/syntax/compiler.py

```diff
@@ -162,6 +259,202 @@ def transformPropertyDef(self, node: s.PropertyDef):
+        # TODO(shun): assert not in behavior or compose
+
+        # --- Extract preconditions and invariants ---
+        preconditions: list[s.Precondition] = []
+        invariants: list[s.Invariant] = []
+        for n in node.header:
+            if isinstance(n, s.Precondition):
+                preconditions.append(n)
+            elif isinstance(n, s.Invariant):
+                invariants.append(n)
+            else:
+                assert (
+                    False
+                ), f"Unexpected node type {n.__class__.__name__} in BehaviorDef header"
+
+        # --- Copy arguments to the behavior object's namespace ---
+
+        # list of all arguments
+        allArgs = itertools.chain(
+            node.args.posonlyargs, node.args.args, node.args.kwonlyargs
+        )
+        # statements that create argument variables
+        copyArgs: list[ast.AST] = []
+        for arg in allArgs:
+            dest = ast.Attribute(
+                ast.Name(behaviorArgName, loadCtx), arg.arg, ast.Store()
+            )
+            copyArgs.append(
+                ast.copy_location(
+                    ast.Assign([dest], ast.Name(arg.arg, loadCtx)), arg
+                )
+            )
+
+        # --- Create a new `arguments` ---
+        newArgs: ast.arguments = self.visit(node.args)
+        # add private current behavior argument and implicit `self` argument
+        newArgs.posonlyargs = initialBehaviorArgs + newArgs.posonlyargs
+
+        # --- Process body ---
+        self.inBehavior = True
+        oldBehaviorLocals = self.behaviorLocals
+
+        self.behaviorLocals = allLocals = LocalFinder.findIn(node.body)
+
+        # handle docstring
+        body = node.body
+        docstring = []
+        if node.docstring is not None:
+            docstring = [ast.Expr(ast.Constant(node.docstring))]
+
+        # process body
+        statements = self.visit(body)
+
+        generatorBody = copyArgs + statements
+        generatorDefinition = ast.FunctionDef(
+            name="makeGenerator",
+            args=newArgs,
+            body=generatorBody,
+            decorator_list=[],
+            returns=None,
+        )
+
+        # --- Save local variables ---
+        saveLocals = ast.Assign(
+            [ast.Name("_locals", ast.Store())], ast.Constant(frozenset(allLocals))
+        )
+
+        # --- Guards ---
+        guardCheckers = self.makeGuardCheckers(newArgs, preconditions, invariants)
+
+        # --- Create class definition ---
+        classBody = docstring + guardCheckers + [saveLocals, generatorDefinition]
+        classDefinition = ast.ClassDef(
+            node.name, [ast.Name("Behavior", loadCtx)], [], classBody, []
+        )
+
+        self.inBehavior = False
+        self.behaviorLocals = oldBehaviorLocals
+
+        return ast.copy_location(classDefinition, node)
+
+    def visit_MonitorDef(self, node: s.MonitorDef):
+        # TODO(shun): assert not in behavior or compose
```

Comment:
I think it would be clearer to unify this with `visit_BehaviorDef`, i.e. break out the code into a common helper function. Once we allow monitors to take arguments, the only difference will be in the superclass used (`Behavior` vs. `Monitor`).

Reply:
Fixed this in b86446e629391f0a45b1c026d4dbc17018825ff9. Created the `makeBehaviorLikeDef` helper function and moved the shared code there.

At this moment, `Behavior` takes two additional arguments: `args` and `header` (preconditions&invariants). When calling the helper from `visit_Monitor`, it uses the empty arguments for `args` and the empty list for the header.